### PR TITLE
✨Bug fixing and improvements

### PR DESCRIPTION
**Pull request by miroslavpejic85**
## Bug fixing and improvements

Hello @vasanthv,

I made some `Bug fixing` and `improvements` to give my contribution:

1. `Bug Fix` https://github.com/vasanthv/talk/issues/36

---

2. `Bug Fix` cam video/icon when screen share

```js
Steps to reproduce:
 - Put the camera off
 - Share a screen
 - Icon remain like a video is off when screen is sharing
 - Stop screen share
 - Cam is back on, but the icon remain off, and not work properly
```

---

3. `Bug Fix` cam video/icon when change camera device

```js
Steps to reproduce:
 - Put the camera off
 - Change the camera device
 - Icon remain like a video is off, and not work properly
```

---

4. `Bug fix` message chat is submitted also if no peers in the channel

```js
Steps to reproduce:
 - No peers in the channel
 - Type chat messages
 - Press ENTER to submit
```

---

5. `Bug fix` missing COPY `postinstall.js` in Dockerfile 

```bash
# docker-compose up
Creating network "talk_default" with the default driver
Building talk
[+] Building 10.0s (8/9)                                                                                                                                                                      
 => [internal] load build definition from Dockerfile                                                                                                                                     0.0s
 => => transferring dockerfile: 160B                                                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                                        0.0s
 => => transferring context: 34B                                                                                                                                                         0.0s
 => [internal] load metadata for docker.io/library/node:latest                                                                                                                           1.2s
 => [1/5] FROM docker.io/library/node:latest@sha256:a6f295c2354992f827693a2603c8b9b5b487db4da0714f5913a917ed588d6d41                                                                     0.0s
 => [internal] load build context                                                                                                                                                        0.0s
 => => transferring context: 3.16kB                                                                                                                                                      0.0s
 => CACHED [2/5] WORKDIR /usr/src/app                                                                                                                                                    0.0s
 => CACHED [3/5] COPY package.json .                                                                                                                                                     0.0s
 => ERROR [4/5] RUN npm install                                                                                                                                                          8.7s
------                                                                                                                                                                                        
 > [4/5] RUN npm install:                                                                                                                                                                     
#8 8.533 
#8 8.533 > talk@2.0.4 postinstall
#8 8.533 > node postinstall.js
#8 8.533 
#8 8.582 node:internal/modules/cjs/loader:959
#8 8.582   throw err;
#8 8.582   ^
#8 8.582 
#8 8.582 Error: Cannot find module '/usr/src/app/postinstall.js'
#8 8.582     at Module._resolveFilename (node:internal/modules/cjs/loader:956:15)
#8 8.582     at Module._load (node:internal/modules/cjs/loader:804:27)
#8 8.582     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
#8 8.582     at node:internal/main/run_main_module:17:47 {
#8 8.582   code: 'MODULE_NOT_FOUND',
#8 8.582   requireStack: []
#8 8.582 }
```
---

6. `Improvements` On server start, add a link in the logs to open directly in browser on click

```js
Server { listening_on: 'http://localhost:3000' }
```

---

<br>

7. `Improvements` on `video/screen off`, show `avatar img` instead of black screen

video-off-ui

---

<br>

8. `As required` how to display the `name` here: https://github.com/vasanthv/talk/issues/9#issuecomment-863066162 I'm added also this functionality

add-peer-name

---

<br>

9. `Improvements` added more `userData` aka peer info

userData-info

---

<br>

9. `Improvements` added `audio status` icon

audio-status

---

<br>

10. `Improvements`, as required: https://github.com/vasanthv/talk/issues/43#issue-1338082841, I'm added the `exit` (hang-up) button, on clicked, you will `leave` the current `room` and `join` in the `new one` (random, as default)

exit-hang-up

---

11. `Updates` updated dependencies (express - socket.io)

```js
	"dependencies": {
		"express": "^4.18.1", 
		"socket.io": "^4.5.1",	
        }	
```

---

12. `Improvements` Not shown share screen button on mobile, not supported at all, ref: 

https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getDisplayMedia#browser_compatibility

---

13. `Improvements` added `use strict` It helps you to write cleaner code, like preventing you from using undeclared variables.

---

14. `Improvements` added `dark theme` (toggleTheme on settings) the `last toggled theme` will be loaded on `app start`.

add-dark-mode

---

15. `Improvements` Reduced the `Docker image` From `1.02GB` to `142MB`

```bash

miroslavpejic@MacBook-Pro-di-Miroslav ~/g/c/talk (master)> docker images
REPOSITORY     TAG       IMAGE ID       CREATED          SIZE
talk_talk      latest    307d674b04de   19 minutes ago   142MB
```
---

16. `Improvements` Added `ngrok`: https://ngrok.com/ (optional) 

If you are set correctly in your `env var` `process.env.NGROK_AUTH_TOKEN` you will see something like this in the Server logs on startup:

```js
> talk@3.0.0 start
> node server.js

Server {
  listen_on: 'http://localhost:3000',
  tunnel_https: 'https://17e3-79-17-23-159.ngrok.io',
  node_version: '16.15.0'
}
```

By clicking on `tunnel_https` URL, Talk will be opened in browser exposed on `https`. So you can use the App directly from your local host, or any other hosts, and be reachable by anyone from outside!

---

<br>

## Notes

The current `Turn server` you using seems to not be reachable & can be the issue of:
- https://github.com/vasanthv/talk/issues/37
- https://github.com/vasanthv/talk/issues/29
- https://github.com/vasanthv/talk/issues/28

---

WebRTC-ice-servers

I Recommend to use only one Stun/Turn and maybe you can install your own as service like [coturn](https://github.com/coturn/coturn) or alternatively, these are reachable ;)

```js
const ICE_SERVERS = [
    { urls: "stun:stun.l.google.com:19302" },
    { urls: "turn:openrelay.metered.ca:443", username: "openrelayproject", credential: "openrelayproject" }
]
```

---

BR. Miroslav Pejic

### CODE REVIEW

**Comment by miroslavpejic85** [comment]
## Add Peer Name 

As required here: https://github.com/vasanthv/talk/issues/9#issuecomment-863066162 added also this functionality

add-peer-name

**Comment by pragmatech** [comment]
While A/B testing with vasanthv's, and miroslavpejic85's PR version, the latter has a bug wherein the control button group vanishes upon clicking on the settings button. (After seeing of vasanthv's laudable appreciation of minimalism, perhaps there are tradeoffs along such a metric. Still, the combo of Miro's chops with this aesthetic is exciting. Thanks!)

**Comment by pragmatech** [comment]
@miroslavpejic85,
Interesting to see your chops added to this relatively light webrtc implementation.
Seeing how, for example, the addition of Ngrok adds @30mb of node modules (which seemingly serves a fairly select purpose i.e. connecting through localhost), my general question: 
Is your fork destined to explore the 'more bells and whistles'-development? Please let me know if I'm missing something on Ngrok's bundle-size payoff.
Thanks for the continual education you provide,
Darren

**Comment by miroslavpejic85** [comment]
> @miroslavpejic85,
> Interesting to see your chops added to this relatively light webrtc implementation.
> Seeing how, for example, the addition of Ngrok adds @30Mb of node modules (which seemingly serves a fairly select purpose i.e. connecting through localhost), my general question:
> Is your fork destined to explore the 'more bells and whistles'-development? Please let me know if I'm missing something on Ngrok's bundle-size payoff.
> Thanks for the continual education you provide,
> Darren

hello @pragmatech,

I wanted to give my contribution to the project as a sign of thanks to @vasanthv, from which I took inspiration about a year ago for a similar project of mine. At the time I didn't have much experience on WebRTC, but over time I acquired it, and I wanted to share my experience to keep this project as simple as it should be, but with some improvements and bug fixing.

As for Ngrok, more information can be found here on what it is: https://ngrok.com/

In order to use it on Talk app you need to do 2 steps:

1. `Sign up for free` and you will receive `YourNgrokAuthToken`.
2. `Set this token` in env var before to start the app like:

```bash
NGROK_AUTH_TOKEN="YourNgrokAuthToken" npm start
```

Or if you want to keep it static just add it on `server.js` here:

```js
const NGROK_AUTH_TOKEN = process.env.NGROK_AUTH_TOKEN || "YourNgrokAuthToken";
```

If set correctly, in the Server log console you should see something like:

```js
> talk@3.0.0 start
> node server.js

Server {
  listen_on: 'http://localhost:3000',
  tunnel_https: 'https://17e3-79-17-23-159.ngrok.io',
  node_version: '16.15.0'
}
```

The `tunnel_https` URL is generated every time you run the app, and with the URL provided, just open the app in the browser, wherever you are from any host and you will be reachable in `HTTPS` by anyone from anywhere ;)

Regarding the sizes, I also improved the docker image, before my tweaks and improvements it weighed `1.02GB` now with the ngrok option included is `203MB`.

```bash
miroslavpejic@MacBook-Pro-di-Miroslav ~> docker images
REPOSITORY     TAG       IMAGE ID       CREATED        SIZE
talk_talk      latest    5f4535ab6825   14 hours ago   203MB
```

I was also pleased to exchange some emails with vasanthv, currently and busy / traveling, as soon as he gets a chance, he will check this PR.

Best Regards.
Miroslav

**Comment by ghost** [comment]
Love this PR. Please someone review and merge it :dancers:

**Comment by vasanthv** [comment]
Ohh! That’s lotta changes. I am gonna go over them this weekend.
Thanks @miroslavpejic85 for your great work.

**Comment by miroslavpejic85** [comment]
> Ohh! That’s lotta changes. I am gonna go over them this weekend.
> Thanks @miroslavpejic85 for your great work.

Take your time, you're welcome! ;)

**Comment by vasanthv** [comment]
@miroslavpejic85  Wow, i looked at the PR and it is great. Neve thought someone will spend this much time to add features to this repo. Thank you so much.

File: www/index.html [hunk 1]
```diff
@@ -27,8 +27,9 @@
 		<link rel="stylesheet" href="/style.css" />
 		<script defer src="https://webrtc.github.io/adapter/adapter-latest.js"></script>
 		<script defer src="/socket.io/socket.io.js"></script>
-		<script defer src="vue.min.js"></script>
+		<script defer src="vue.min.js"></script> <!-- TODO where is? -->
```
Comment (vasanthv) [inline www/index.html:30]: This comes from https://github.com/vasanthv/talk/blob/master/server.js#L21

**Comment by miroslavpejic85** [comment]
> @miroslavpejic85 Wow, i looked at the PR and it is great. Neve thought someone will spend this much time to add features to this repo. Thank you so much.

It's a pleasure, i'm glad you like it.

**Pull request merged**